rpc: combinerawtransaction now rejects unmergeable transactions

adamandrews1 opened the pull request:
Previously, combinerawtransaction would silently return the first tx when asked to combine unrelated txs. Now, it will check tx mergeability and throws a descriptive error if tx cannot be merged.

fixes  #25980

adamandrews1 committed:
rpc: combinerawtransaction now rejects unmergeable transactions

Previously, combinerawtransaction would silently return the first tx when
asked to combine unrelated txs. Now, it will check tx mergeability and
throws a descriptive error if tx cannot be merged.

adamandrews1 commented:
It seems like, with https://github.com/bitcoin/bitcoin/pull/31249, I should migrate these test changes as well. WIP

adamandrews1 commented:
> Can this use the same approach and error message as `combinepsbt`?

This is the same functional approach and error message already.
Creating a PSBT involves stripping the scriptSig and scriptWitnesses. Later, the hashes are compared.  That's exactly what this change is doing too but the syntax is different. I can align the syntax to make it more clear it is doing the same thing.

1440000bytes reviewed (approved):
utACK https://github.com/bitcoin/bitcoin/pull/31298/commits/85c216871a01aab003e1b1aba6246b4178afc929

yuvicc reviewed (approved):
Tested ACK 85c216871a01aab003e1b1aba6246b4178afc929

Below is my full test setup:

## Tested on Regtest mode

### master@d2136d32bb4764e4035fb892d979c27f037fad93
- When checked with only 1 raw txn, returns only the txn (no error for missing txns)
- When checked for mergeability: 
     - returns only the first txn when unrelated txns were provided
     - doesn't show descriptive error for any mergeability issue
     

### combinerawtransaction-check@85c216871a01aab003e1b1aba6246b4178afc929
- When only 1 raw txn provided it returns missing inputs
- checking for mergeability:
     - shows descriptive error for mergeability issue 
for e.g.

```
error code: -8
error message:
Transaction 1 not compatible (different transactions)
```

Gives combine hex when same txns are provided
```
./build/src/bitcoin-cli combinerawtransaction '["0200000001a0d2e799d40d404da2daf128fc678fdd15f245bdd92c6e1ad15b45f7258a47b90000000000fdffffff01008c86470000000016001483f259c98962117d75154f35cb55cd60b88a257e00000000","0200000001a0d2e799d40d404da2daf128fc678fdd15f245bdd92c6e1ad15b45f7258a47b90000000000fdffffff01008c86470000000016001483f259c98962117d75154f35cb55cd60b88a257e00000000"]'
0200000001a0d2e799d40d404da2daf128fc678fdd15f245bdd92c6e1ad15b45f7258a47b90000000000fdffffff01008c86470000000016001483f259c98962117d75154f35cb55cd60b88a257e00000000
```

for inputs already spent:
```
error code: -25
error message:
Input not found or already spent
```

*Note: newly created wallet and txns were used during the test using ```createrawtransaction``` rpc in regtest mode* 

Also, the functional test LGTM:
```
Temporary test directory at /var/folders/jb/wlbrz0t95vl58wzxjt75wqmh0000gn/T/test_runner_₿_🏃_20241219_004940
Remaining jobs: [rpc_createmultisig.py]
1/1 - rpc_createmultisig.py passed, Duration: 3 s

TEST                  | STATUS    | DURATION

rpc_createmultisig.py | ✓ Passed  | 3 s

ALL                   | ✓ Passed  | 3 s (accumulated) 
Runtime: 3 s
```

danielabrozzoni commented on src/rpc/rawtransaction.cpp:
```diff
@@ -650,16 +650,37 @@ static RPCHelpMan combinerawtransaction()
+            // Remove all scriptSigs and scriptWitnesses from inputs
+            for (CTxIn& input : txVariantsCopy[k].vin) {
+                input.scriptSig.clear();
+                input.scriptWitness.SetNull();
```
I don't think it's needed to clear the script witness, as it's not taken into account when calculating the txid

adamandrews1 replied:
This is done to align the method with PSBT. It could be removed without breaking the change. But it also is just clearing data which should take minimal time. I think it's worth keeping to align with PSBT format.

brunoerg replied:
No objections on it but maybe you could update the documentation, it says: "This involves stripping sigscripts from each tx and ensuring txid is consistent." but you're also stripping the script witness.

adamandrews1 replied:
@brunoerg I think you are right, I will update the description here to make it explicit we are tracking PSBT here 👍🏻

zaidmstrr commented on src/rpc/rawtransaction.cpp:
```diff
@@ -639,16 +639,37 @@ static RPCHelpMan combinerawtransaction()
     std::vector<CMutableTransaction> txVariants(txs.size());
 
-    for (unsigned int idx = 0; idx < txs.size(); idx++) {
+    for (unsigned int idx {0}; idx < txs.size(); idx++) {
```
```suggestion
    for (unsigned int idx{0}; idx < txs.size(); idx++) {
```
nit: don't need to add the extra space

i-am-yuvi replied:
It might be helpful to consider giving a `Concept ACK` or `NACK` before the review. I came across this idea [here](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#code-review).

zaidmstrr replied:
Done

adamandrews1 replied:
Fixed nit

zaidmstrr commented on src/rpc/rawtransaction.cpp:
```diff
@@ -639,16 +639,37 @@ static RPCHelpMan combinerawtransaction()
+    { // Test Tx relation for mergeability. This involves converting each tx to PSBT format (stripping sigscripts/witnesses) and ensuring txid is consistent.
+        std::vector<CMutableTransaction> txVariantsCopy(txVariants);
+        Txid txid{};
+        for (unsigned int k{0}; k < txVariantsCopy.size(); k++) {
```
```suggestion
        for (unsigned int k{0}; k < txVariantsCopy.size(); ++k) {
```
nit: `++k` should be given more prefrence then `k++ `

adamandrews1 replied:
Fixed nit

zaidmstrr commented on src/rpc/rawtransaction.cpp:
```diff
@@ -678,7 +699,7 @@ static RPCHelpMan combinerawtransaction()
     const CTransaction txConst(mergedTx);
     // Sign what we can:
-    for (unsigned int i = 0; i < mergedTx.vin.size(); i++) {
+    for (unsigned int i{0}; i < mergedTx.vin.size(); i++) {
```
```suggestion
    for (unsigned int i{0}; i < mergedTx.vin.size(); ++i) {
```
nit: same for here `++i` should be prioritised.

adamandrews1 replied:
Fixed nit

zaidmstrr commented:
Concept ACK

naiyoma reviewed (commented):
Tested ACK https://github.com/bitcoin/bitcoin/pull/31298/commits/9d31e94bf584e12269683b74b1c96c7ac883a760

I have tested the changes on regtest and got expected output 

```
bitcoin-cli -regtest combinerawtransaction "[\"$RAW_TX1\", \"$RAW_TX2\"]"
error code: -8
error message:
Transaction 1 not compatible (different transactions)
```

naiyoma commented on src/rpc/rawtransaction.cpp:
```diff
@@ -639,16 +639,37 @@ static RPCHelpMan combinerawtransaction()
+
+    // Can't merge < 2 items
+    if (txs.size() < 2) {
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Missing transactions");
```
nit: IMO, this could be a bit more descriptive

 ```diff
    // Can't merge < 2 items
     if (txs.size() < 2) {
-        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Missing transactions");
+        throw JSONRPCError(
+            RPC_INVALID_PARAMETER,
+            txs.size() == 0 
+                ? "Missing transactions. At least two transactions required." 
+                : "Insufficient transactions. At least two transactions  required."
+        );
     }
```

rkrux replied:
Nit: Can consider to be more verbose in the error message.

```diff
- throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Missing transactions");
+ throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Missing transactions. At least two transactions required.");
```

yuvicc reviewed (commented):
re-ACK 9d31e94bf584e12269683b74b1c96c7ac883a760

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -639,16 +639,37 @@ static RPCHelpMan combinerawtransaction()
 
-    if (txVariants.empty()) {
-        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Missing transactions");
+    { // Test Tx relation for mergeability. This involves converting each tx to PSBT format (stripping sigscripts/witnesses) and ensuring txid is consistent.
```
This comment is incorrect, it's not doing anything related to PSBT. You can just say that the scriptSigs and scriptWitnesses are being stripped in order to do a txid comparison.

adamandrews1 replied:
Thanks, I will update this comment

theStack reviewed (commented):
Concept ACK

rkrux commented on src/rpc/rawtransaction.cpp:
```diff
@@ -647,16 +647,37 @@ static RPCHelpMan combinerawtransaction()
+                txid = txVariantsCopy[k].GetHash();
+            } else {
+                if (txid != txVariantsCopy[k].GetHash()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Transaction %d not compatible (different transactions)", k));
```
Nit: Reading `Transaction 1` can be confusing for the user if they are not aware that 0-based indexing is being used.

rkrux reviewed (commented):
Concept ACK a26e9fd5b5908853d37b81915f5e12a5ad822f86

Thanks for adding the fix.

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -647,16 +647,37 @@ static RPCHelpMan combinerawtransaction()
     std::vector<CMutableTransaction> txVariants(txs.size());
 
-    for (unsigned int idx = 0; idx < txs.size(); idx++) {
+    for (unsigned int idx{0}; idx < txs.size(); ++idx) {
```
nit: This is an unnecessary style change.

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -686,7 +707,7 @@ static RPCHelpMan combinerawtransaction()
     const CTransaction txConst(mergedTx);
     // Sign what we can:
-    for (unsigned int i = 0; i < mergedTx.vin.size(); i++) {
+    for (unsigned int i{0}; i < mergedTx.vin.size(); ++i) {
```
nit: This is an unnecessary style change.

rkrux commented on test/functional/rpc_createmultisig.py:
```diff
@@ -148,8 +148,38 @@ def do_multisig(self, nkeys, nsigs, output_type):
         assert_raises_rpc_error(-22, "TX decode failed", node2.combinerawtransaction, [rawtx2['hex'], rawtx3['hex'] + "00"])
         assert_raises_rpc_error(-22, "Missing transactions", node2.combinerawtransaction, [])
-        combined_rawtx = node2.combinerawtransaction([rawtx2["hex"], rawtx3["hex"]])
+        assert_raises_rpc_error(-22, "Missing transactions", node2.combinerawtransaction, [rawtx2['hex']])
```
Let's assert the full error for thoroughness.

```diff
-        assert_raises_rpc_error(-22, "Missing transactions", node2.combinerawtransaction, [])
-        assert_raises_rpc_error(-22, "Missing transactions", node2.combinerawtransaction, [rawtx2['hex']])
+        assert_raises_rpc_error(-22, "Missing transactions. At least two transactions required.", node2.combinerawtransaction, [])
+        assert_raises_rpc_error(-22, "Missing transactions. At least two transactions required.", node2.combinerawtransaction, [rawtx2['hex']])
```

rkrux commented on test/functional/rpc_createmultisig.py:
```diff
@@ -148,8 +148,38 @@ def do_multisig(self, nkeys, nsigs, output_type):
+        assert_raises_rpc_error(-8, "Transaction number 2 not compatible with first transaction", node0.combinerawtransaction, [rawtx2['hex'], unrelated_tx])
+
+        # similar transaction but with different locktime, transactions are not mergeable
+        unrelated_tx = node2.createrawtransaction([{"txid": tx["txid"], "vout": tx["sent_vout"]}], [{out_addr: outval}], 1)
```
This is a nice suggestion that can be implemented.

rkrux reviewed (changes requested):
Overall looking quite good.

I noticed that these new cases are added within the `do_multisig` function that is called several times here:
https://github.com/bitcoin/bitcoin/blob/56ce78d5f62c9eb9353d33eedd15495c1205465f/test/functional/rpc_createmultisig.py#L48-L51

Given the nature of the bug regarding the un-mergeable transactions, I feel testing this once should be sufficient in which case we can consider adding these cases in a separate function that should help in avoiding testing the fix multiple times. WDYT? (I have not formed a strong opinion on this though.)

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -614,8 +620,23 @@ static RPCHelpMan combinerawtransaction()
+            if (k == 0) {
+                txid = tx_variants_copy[k].GetHash();
+            } else {
+                if (txid != tx_variants_copy[k].GetHash()) {
```
In a516fe8ada1880f30c6e4553677a9a2580a17cb0 "rpc: combinerawtransaction now rejects unmergeable transactions"

nit: can be a single line `else if`

rkrux reviewed (commented):
> The tx mergeability constraints are invariant across output_type and n_sig etc, it should be sufficient to test them one time explicitly and disable during all other runs. Probably not a huge runtime cost as-is but there is definitely a fan-out that would adds additional overhead to all callers of do_multisig so got rid of that. Also integrated all other feedback.

Thanks for addressing all the feedback! I had a different approach in mind that creates a separate function for these checks but the latest implementation also looks fine to me.

Code is lgtm e3069ecd814db2ece640ba0427e7778b03f7314e

Shared 2 nits that could be addressed at the moment I believe.

sedited commented:
> Sure, I can address.

Pinging again since it has been a month. Just need to apply the two small diffs above by the looks of it.

adamandrews1 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -614,8 +620,21 @@ static RPCHelpMan combinerawtransaction()
+            }
+            if (k == 0) {
+                txid = tx_variants_copy[k].GetHash();
+            } else if (txid != tx_variants_copy[k].GetHash()) {
```
renamed

adamandrews1 commented on test/functional/rpc_createmultisig.py:
```diff
@@ -146,10 +152,32 @@ def do_multisig(self, nkeys, nsigs, output_type):
+            out_addr2 = getnewdestination('bech32')[2]
+            unmergeable_transaction_args = [
+                ([[{"txid": tx["txid"], "vout": tx["sent_vout"]}], [{out_addr: outval * 2}]], {}), # similar transaction but with different amount to send
+                ([[{"txid": tx["txid"], "vout": tx["sent_vout"]}], [{out_addr: outval}]], {"locktime": 0, "replaceable": True, "version": 1}), # similar transaction but with different version
```
removed

adamandrews1 commented:
> > The tx mergeability constraints are invariant across output_type and n_sig etc, it should be sufficient to test them one time explicitly and disable during all other runs. Probably not a huge runtime cost as-is but there is definitely a fan-out that would adds additional overhead to all callers of do_multisig so got rid of that. Also integrated all other feedback.
> 
> Thanks for addressing all the feedback! I had a different approach in mind that creates a separate function for these checks but the latest implementation also looks fine to me.
> 
> Code is lgtm [e3069ec](https://github.com/bitcoin/bitcoin/commit/e3069ecd814db2ece640ba0427e7778b03f7314e)
> 
> Shared 2 nits that could be addressed at the moment I believe.

addressed 2 nits @rkrux @sedited

fanquake commented:
You'll need to squash your commits / remove the merge commit.

adamandrews1 commented:
> You'll need to squash your commits / remove the merge commit.

squashed and rebased

rkrux reviewed (approved):
ACK [6d86184](https://github.com/bitcoin/bitcoin/pull/31298/commits/6d86184a8bcc34186d217cc7cfbf7acc21af7ca4)

Thanks for addressing the nits.

nervana21 commented:
Concept ACK

nervana21 commented:
tACK 6d86184a8bcc34186d217cc7cfbf7acc21af7ca4

Small nit: I tend to agree with other reviewers who state that clearing the `scriptWitness` even though it's not necessary makes the code harder to reason about

achow101 commented:
ACK 6d86184a8bcc34186d217cc7cfbf7acc21af7ca4
